File: admin/internal/domain/table_approval.go
package domain

import (
	"context"
	"errors"
)

var ErrTableApprovalVersion = errors.New("table approval assignment changed")

type ApprovalRoleIdentity struct {
	ID   string `json:"id"`
	Name string `json:"name"`
}

type TableApprovalAssignment struct {
	TableName string                 `json:"table_name"`
	Version   string                 `json:"version"`
	RoleIDs   []string               `json:"role_ids"`
	Roles     []ApprovalRoleIdentity `json:"roles"`
}

type TableApprovalChange struct {
	ActorID, TableName, ExpectedVersion, RequestKey string
	RoleIDs                                         []string
}

type TableApprovalRepository interface {
	ReadTableApproval(context.Context, string) (TableApprovalAssignment, error)
	SaveTableApproval(context.Context, TableApprovalChange) (TableApprovalAssignment, error)
}

File: web/src/api/table-approval.ts
import { z } from "zod";
import { request } from "./client";

export const approvalRoleIdentitySchema = z.object({ id: z.string(), name: z.string() });
const assignmentSchema = z.object({ table_name: z.string(), version: z.string().regex(/^(0|[1-9][0-9]*)$/), role_ids: z.array(z.string()), roles: z.array(approvalRoleIdentitySchema) });
export type TableApprovalAssignment = z.infer<typeof assignmentSchema>;
export type ApprovalRoleIdentity = z.infer<typeof approvalRoleIdentitySchema>;
export type TableApprovalInput = { expected_version: string; role_ids: string[] };
export const tableApprovals = {
  get: (table: string) => request(`/api/v1/table-policies/${encodeURIComponent(table)}/approval-roles`, { schema: assignmentSchema }),
  save: (table: string, input: TableApprovalInput, key: string) => request(`/api/v1/table-policies/${encodeURIComponent(table)}/approval-roles`, { method: "PUT", body: JSON.stringify(input), headers: { "Idempotency-Key": key }, schema: assignmentSchema }),
};
